---
title: "ggheat : Heatmaps with ggplot"
author: "Ira R. Cooke"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo=FALSE,warning=FALSE,include=FALSE}
  library(ggplotheatmap, quietly = TRUE, warn.conflicts = TRUE)
```


Heatmaps are a visualisation tool that combine a spatial arrangement of data points (the `map`) with a colour scale (the `heat`) indicating the measured value at each point. They are particularly useful for gene expression data where the expression value is measured for many genes across many samples. Typically a gene expression heatmap is a grid, with rows representing genes and columns representing samples (or vice versa). The spatial arrangement in this context is achieved through clustering rows and/or columns in order to place genes (or samples) with similar expression values closer to each other.

The `ggheat` package is an extension of the `ggplot2` framework for creating gene expression style heatmaps. While there are many excellent existing heatmap packages for `R` they are all standalone in the sense that they are built on lower level graphics frameworks such as `grid` or `lattice` rather than on a higher level plotting framework such as `ggplot2`.  By creating a heatmap extension for `ggplot`, `ggheat` automatically inherits many powerful features from its parent package including;

  - the ability to change the way points are plotted (eg a point, filled rectangle, arrow)
  - the ability to overlay additional contextual data using an appropriate geometric representation
  - the ability to apply the statistical transformation (clustering) to groups of data separately
  - complete control over the appearance of the plot (eg font, colour, scales etc) using a framework that is familiar to many users and is already very well documented.

# Getting started

To illustrate `ggheat` we start by generating some data.

```{r}
  d <- sapply(c(1,1,1,2,2,2),function(rx) { 
    c(sapply(c(1,2),function(cy) rnorm(10,mean=rx*cy,sd = 0.2))) 
    })
```

This produces a matrix with 20 rows and 6 columns.  This kind of object can be passed directly to ggheat.

```{r}
  gh <- ggheat(d)
```

This produces the object `gh` which contains our data in a form that is ready to plot. To actually create a plot we must add a plottable layer.

```{r, fig.show='hold'}
  gh + stat_heat()
```

Note that we used the `+` operator to add a layer in a way that works just like `ggplot`. One of the key principles of `ggplot` is that every layer must have both a stat and a geom. For heatmaps the stat or statistical transformation is a clustering which rearranges rows and/or columns.  The default geom for this stat is the `heat` geom which is a slightly modified version of ggplot's `geom_tile` that deals with non-numeric rows and columns properly.  

Although `stat_heat` is a bit special we can also apply other standard ggplot functions to further modify the plot.  For example to swap rows and columns we can use `coord_flip`.

```{r, fig.show='hold'}
  gh + stat_heat() + coord_flip()
```




# Advanced


The ggheat package comes with an example dataset from a study by Caruana et al (cite). The dataset includes measurements of mRNA expression in various tissues of a charismatic marine animal (the Southern Bottletail Squid, see figure 1).  

Load and inspect the built in dataset


```{r}
  data(squid)

  head(squid,n=2)
```

The mRNA expression measurements are labelled according to their tissue type. They are `Arms`, `Brain`, `Slime`, `VM`, and `DM`.  The latter two are abbreviations for `Ventral Mantle` and `Dorsal Mantle` respectively.  In addition to these five columns, the data also contains three other columns, `iBAQ` which is a mass spectrometry based measure of protein abundance within the `Slime`, `prot_id` which lists the coding orfs for this transcript and `protein_group` which groups transcripts into according to whether they can be distinguished at the protein level.

The first step is to create a `ggheat` object.  We supply the `id.vars` argument in order to identify the columns that do not contain values to be used to create the heatmap. 

```{r}
  gh <- ggheat(squid, id.vars = colnames(squid[,1:3]))
```


Notice that

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
